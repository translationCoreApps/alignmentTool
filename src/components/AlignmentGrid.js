import React, {Component} from 'react';
import PropTypes from 'prop-types';
// constants
import * as types from './WordCard/Types';
// components
import AlignmentCard from './AlignmentCard';

const makeStyles = props => {
  return {
    root: {
      display: 'flex',
      flexWrap: 'wrap',
      backgroundColor: '#ffffff',
      padding: '0px 10px 10px',
      overflowY: 'auto',
      flexGrow: 2,
      direction: props.sourceDirection,
      alignContent: 'flex-start'
    },
    warning: {
      padding: '20px',
      backgroundColor: '#ccc',
      display: 'inline-block'
    }
  };
};

/**
 * Renders a grid of word/phrase alignments
 */
class AlignmentGrid extends Component {
  render() {
    const {
      translate,
      actions,
      lexicons,
      onCancelSuggestion,
      sourceDirection,
      targetDirection,
      onAcceptTokenSuggestion,
      alignments,
      contextId
    } = this.props;

    if (!contextId) {
      return <div/>;
    }

    const styles = makeStyles(this.props);

    // TODO: add support for dragging to left of card. See utils/dragDrop.js
    return (
      <div id='AlignmentGrid' style={styles.root}>
        {
          alignments.map((alignment, key) => {
            return (
              <React.Fragment key={key}>
                {/* placeholder for un-merging primary words */}
                {/* TODO: cannot place this here due to this bug https://github.com/react-dnd/react-dnd/issues/735*/}
                {/*<AlignmentCard*/}
                {/*translate={translate}*/}
                {/*alignmentIndex={index}*/}
                {/*placeholderPosition="left"*/}
                {/*bottomWords={[]}*/}
                {/*topWords={[]}*/}
                {/*onDrop={item => this.handleDrop(index, item)}*/}
                {/*actions={actions}*/}
                {/*lexicons={lexicons}*/}
                {/*/>*/}

                <AlignmentCard
                  translate={translate}
                  sourceDirection={sourceDirection}
                  targetDirection={targetDirection}
                  onCancelTokenSuggestion={onCancelSuggestion}
                  onAcceptTokenSuggestion={onAcceptTokenSuggestion}
                  alignmentIndex={alignment.index}
                  isSuggestion={alignment.isSuggestion}
                  targetNgram={alignment.targetNgram}
                  sourceNgram={alignment.sourceNgram}
                  onDrop={item => this.handleDrop(alignment.index, item)}
                  actions={actions}
                  lexicons={lexicons}
                />
                {/* placeholder for un-merging primary words */}
                <AlignmentCard
                  translate={translate}
                  sourceDirection={sourceDirection}
                  targetDirection={targetDirection}
                  alignmentIndex={alignment.index}
                  isSuggestion={alignment.isSuggestion}
                  placeholderPosition="right"
                  targetNgram={[]}
                  sourceNgram={[]}
                  onDrop={item => this.handleDrop(alignment.index, item)}
                  actions={actions}
                  lexicons={lexicons}
                />
              </React.Fragment>
            );
          })
        }
      </div>
    );
  }

  handleDrop(alignmentIndex, item) {
    const {onDropTargetToken, onDropSourceToken} = this.props;
    if (item.type === types.SECONDARY_WORD) {
      if (item.tokens) {
        // drop selected tokens
        for (let i = 0; i < item.tokens.length; i++) {
          onDropTargetToken(item.tokens[i], alignmentIndex, item.alignmentIndex);
        }
      } else {
        // drop single token
        onDropTargetToken(item.token, alignmentIndex, item.alignmentIndex);
      }
    }
    if (item.type === types.PRIMARY_WORD) {
      onDropSourceToken(item.token, alignmentIndex, item.alignmentIndex);
    }
  }
}

AlignmentGrid.propTypes = {
  onDropTargetToken: PropTypes.func.isRequired,
  onDropSourceToken: PropTypes.func.isRequired,
  onCancelSuggestion: PropTypes.func.isRequired,
  onAcceptTokenSuggestion: PropTypes.func.isRequired,
  alignments: PropTypes.array.isRequired,
  contextId: PropTypes.object,
  translate: PropTypes.func.isRequired,
  actions: PropTypes.object.isRequired,
  lexicons: PropTypes.object.isRequired,
  sourceDirection: PropTypes.oneOf(['ltr', 'rtl']),
  targetDirection: PropTypes.oneOf(['ltr', 'rtl'])
};

AlignmentGrid.defaultProps = {
  sourceDirection: 'ltr',
  targetDirection: 'ltr'
};

export default AlignmentGrid;
