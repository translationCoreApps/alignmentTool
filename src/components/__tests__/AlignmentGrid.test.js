import React from 'react';
import AlignmentGrid from '../AlignmentGrid';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import TestBackend from 'react-dnd-test-backend';
import {DragDropContext} from 'react-dnd';
import {Token} from 'wordmap-lexer';

test('empty snapshot', () => {
  const wrapper = renderer.create(
    <AlignmentGrid lexicons={{}}
                   alignments={[]}
                   onAcceptTokenSuggestion={jest.fn()}
                   onDropSourceToken={jest.fn()}
                   onCancelSuggestion={jest.fn()}
                   onDropTargetToken={jest.fn()}
                   translate={k => k}
                   actions={{
                     showPopover: jest.fn(),
                     loadLexiconEntry: jest.fn(),
                     getLexiconData: jest.fn()
                   }}/>
  );
  expect(wrapper).toMatchSnapshot();
});

describe('AlignmentGrid', () => {
  let contextId, alignmentData;

  beforeEach(() => {

    contextId = {
      'groupId': 'figs_metaphor',
      'occurrence': 1,
      'quote': 'that he put before them',
      'information': 'Paul speaks about good deeds as if they were objects that God could place in front of people. AT: "that God prepared for them to do" (See: [[:en:ta:vol1:translate:figs_metaphor]]) \n',
      'reference': {
        'bookId': 'luk',
        'chapter': 1,
        'verse': 1
      },
      'tool': 'TranslationNotesChecker'
    };
    alignmentData = {
      '1': {
        '1': [
          {
            'position': 0,
            'index': 0,
            'isSuggestion': false,
            'sourceNgram': [
              new Token({
                'text': 'ἐπειδήπερ',
                'strongs': 'G18950',
                'lemma': 'ἐπειδήπερ',
                'morph': 'Gr,CS,,,,,,,,',
                'occurrence': 1,
                'occurrences': 1
              })
            ],
            'targetNgram': []
          },
          {
            'position': 1,
            'index': 1,
            'isSuggestion': false,
            'sourceNgram': [
              new Token({
                'text': 'πολλοὶ',
                'strongs': 'G41830',
                'lemma': 'πολλός',
                'morph': 'Gr,RI,,,,NMP,',
                'occurrence': 1,
                'occurrences': 1
              })
            ],
            'targetNgram': []
          },
          {
            'position': 2,
            'index': 2,
            'isSuggestion': false,
            'sourceNgram': [
              new Token({
                'text': 'ἐπεχείρησαν',
                'strongs': 'G20210',
                'lemma': 'ἐπιχειρέω',
                'morph': 'Gr,V,IAA3,,P,',
                'occurrence': 1,
                'occurrences': 1
              })
            ],
            'targetNgram': []
          }
        ]
      }
    };
  });

  test('renders Luke 1:1', () => {
    // given
    contextId.reference.chapter = '1';
    contextId.reference.verse = '1';
    const ConnectedAlignmentGrid = wrapInTestContext(AlignmentGrid);
    const wrapper = renderer.create(
      <ConnectedAlignmentGrid
        alignments={alignmentData['1']['1']}
        contextId={contextId}
        onAcceptTokenSuggestion={jest.fn()}
        onDropSourceToken={jest.fn()}
        onDropTargetToken={jest.fn()}
        onCancelSuggestion={jest.fn()}
        translate={k => k}
        actions={{
          showPopover: jest.fn(),
          loadLexiconEntry: jest.fn(),
          getLexiconData: jest.fn()
        }}
        lexicons={{}}
      />
    );

    // then
    expect(wrapper).toMatchSnapshot();
  });

  test('renders undefined Luke 1:81 without crashing', () => {
    // given
    const chapter = '1';
    const verse = '81';
    contextId.reference.chapter = chapter;
    contextId.reference.verse = verse;
    const expectedWords = 0;

    // when
    const enzymeWrapper = shallow(
      <AlignmentGrid
        onAcceptTokenSuggestion={jest.fn()}
        alignments={alignmentData['1']['1']}
        contextId={contextId}
        onDropSourceToken={jest.fn()}
        onCancelSuggestion={jest.fn()}
        onDropTargetToken={jest.fn()}
        translate={k => k}
        actions={{
          showPopover: jest.fn(),
          loadLexiconEntry: jest.fn(),
          getLexiconData: jest.fn()
        }}
        lexicons={{}}
      />
    );

    // then
    const dropBoxArea = enzymeWrapper.find('div');
    expect(dropBoxArea).toBeTruthy();
    expect(dropBoxArea.getElements().length).toEqual(expectedWords + 1);
  });
});

/**
 * Wraps a component into a DragDropContext that uses the TestBackend.
 */
function wrapInTestContext(DecoratedComponent) {
  return DragDropContext(TestBackend)(
    class TestContextContainer extends React.Component {
      render() {
        return <DecoratedComponent {...this.props} />;
      }
    }
  );
}
