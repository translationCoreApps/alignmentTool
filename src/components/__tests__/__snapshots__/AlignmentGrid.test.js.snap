// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlignmentGrid renders Luke 1:1 1`] = `
<div
  id="AlignmentGrid"
  style={
    Object {
      "alignContent": "flex-start",
      "backgroundColor": "#ffffff",
      "display": "flex",
      "flexGrow": 2,
      "flexWrap": "wrap",
      "overflowY": "auto",
      "padding": "0px 10px 10px",
    }
  }
>
  <div>
    <div
      style={
        Object {
          "backgroundColor": "#DCDCDC",
          "flexGrow": 0,
          "margin": "0px 10px 10px 0px",
          "minWidth": "115px",
          "padding": "7px",
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "border": "3px dashed transparent",
              "boxSizing": "border-box",
              "flexGrow": 1,
              "marginBottom": "7px",
              "minHeight": "45px",
              "transition": "0.1s",
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "display": "flex",
                "left": 0,
                "opacity": 1,
                "position": "relative",
                "top": 0,
                "transition": "0.1s",
              }
            }
          >
            <div
              onClick={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
              style={
                Object {
                  "flex": 1,
                  "position": "relative",
                }
              }
            >
              <div>
                <div
                  style={
                    Object {
                      "flex": 1,
                    }
                  }
                >
                  <div
                    style={
                      Object {
                        "backgroundColor": "#333333",
                        "borderLeft": "5px solid #44C6FF",
                        "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset",
                        "color": "#ffffff",
                        "cursor": "pointer",
                        "display": "flex",
                        "flexDirection": "row",
                        "padding": "9px",
                      }
                    }
                  >
                    <span
                      style={
                        Object {
                          "display": "flex",
                          "flex": 1,
                        }
                      }
                    >
                      <span
                        onClick={[Function]}
                        style={
                          Object {
                            "flexGrow": 2,
                          }
                        }
                      >
                        ἐπειδήπερ
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "border": "3px dashed #ffffff",
              "boxSizing": "border-box",
              "flexGrow": 1,
              "minHeight": "45px",
              "transition": "0.1s",
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "display": "flex",
                "left": 0,
                "opacity": 1,
                "position": "relative",
                "top": 0,
                "transition": "0.1s",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "display": "none",
        "transition": "0.5s",
        "width": "0",
      }
    }
  />
  <div>
    <div
      style={
        Object {
          "backgroundColor": "#DCDCDC",
          "flexGrow": 0,
          "margin": "0px 10px 10px 0px",
          "minWidth": "115px",
          "padding": "7px",
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "border": "3px dashed transparent",
              "boxSizing": "border-box",
              "flexGrow": 1,
              "marginBottom": "7px",
              "minHeight": "45px",
              "transition": "0.1s",
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "display": "flex",
                "left": 0,
                "opacity": 1,
                "position": "relative",
                "top": 0,
                "transition": "0.1s",
              }
            }
          >
            <div
              onClick={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
              style={
                Object {
                  "flex": 1,
                  "position": "relative",
                }
              }
            >
              <div>
                <div
                  style={
                    Object {
                      "flex": 1,
                    }
                  }
                >
                  <div
                    style={
                      Object {
                        "backgroundColor": "#333333",
                        "borderLeft": "5px solid #44C6FF",
                        "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset",
                        "color": "#ffffff",
                        "cursor": "pointer",
                        "display": "flex",
                        "flexDirection": "row",
                        "padding": "9px",
                      }
                    }
                  >
                    <span
                      style={
                        Object {
                          "display": "flex",
                          "flex": 1,
                        }
                      }
                    >
                      <span
                        onClick={[Function]}
                        style={
                          Object {
                            "flexGrow": 2,
                          }
                        }
                      >
                        πολλοὶ
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "border": "3px dashed #ffffff",
              "boxSizing": "border-box",
              "flexGrow": 1,
              "minHeight": "45px",
              "transition": "0.1s",
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "display": "flex",
                "left": 0,
                "opacity": 1,
                "position": "relative",
                "top": 0,
                "transition": "0.1s",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "display": "none",
        "transition": "0.5s",
        "width": "0",
      }
    }
  />
  <div>
    <div
      style={
        Object {
          "backgroundColor": "#DCDCDC",
          "flexGrow": 0,
          "margin": "0px 10px 10px 0px",
          "minWidth": "115px",
          "padding": "7px",
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "border": "3px dashed transparent",
              "boxSizing": "border-box",
              "flexGrow": 1,
              "marginBottom": "7px",
              "minHeight": "45px",
              "transition": "0.1s",
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "display": "flex",
                "left": 0,
                "opacity": 1,
                "position": "relative",
                "top": 0,
                "transition": "0.1s",
              }
            }
          >
            <div
              onClick={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
              style={
                Object {
                  "flex": 1,
                  "position": "relative",
                }
              }
            >
              <div>
                <div
                  style={
                    Object {
                      "flex": 1,
                    }
                  }
                >
                  <div
                    style={
                      Object {
                        "backgroundColor": "#333333",
                        "borderLeft": "5px solid #44C6FF",
                        "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset",
                        "color": "#ffffff",
                        "cursor": "pointer",
                        "display": "flex",
                        "flexDirection": "row",
                        "padding": "9px",
                      }
                    }
                  >
                    <span
                      style={
                        Object {
                          "display": "flex",
                          "flex": 1,
                        }
                      }
                    >
                      <span
                        onClick={[Function]}
                        style={
                          Object {
                            "flexGrow": 2,
                          }
                        }
                      >
                        ἐπεχείρησαν
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "border": "3px dashed #ffffff",
              "boxSizing": "border-box",
              "flexGrow": 1,
              "minHeight": "45px",
              "transition": "0.1s",
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "display": "flex",
                "left": 0,
                "opacity": 1,
                "position": "relative",
                "top": 0,
                "transition": "0.1s",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "display": "none",
        "transition": "0.5s",
        "width": "0",
      }
    }
  />
</div>
`;

exports[`empty snapshot 1`] = `<div />`;
